### DISCUSSION TITLE
Double activation: get machine ID

### DISCUSSION THREAD

**Discussion by nschloe** [keygen-sh/community#65] [Q&A] [answered]
When activating a machine for a license, the API will return the machine ID:

```python
{
    'data': {
        'id': '...',
        'type': 'machines',
        'attributes': {
            'fingerprint': '...',
             # ...
}
```

When trying to activate the same machine a second time, the call fails with

```python
{
    'meta': {'id': '...'},
    'errors': [
        {
            'title': 'Unprocessable resource',
            'detail': 'has already been taken',
            'code': 'FINGERPRINT_TAKEN',
            # ...
        }
    ]
}
```

I don't know why this is an error. Will it also error out if the same machine tries to activate a different license (with the same machine fingerprint)?

Also, what's the `meta.id` here? I had assumed the machine ID, but apparently it's not. Do I have to search machines by fingerprint to get the ID?

**Comment by ezekg** [comment, marked as answer]
You're receiving an error because you can't activate a machine that is already activated. Typically, if you're doing this, then your activation flow is incorrect. Please follow [this activation flow](https://keygen.sh/docs/activating-machines/), i.e. you should validate the license *before* activating the current machine, to check if it's already activated and valid. If it's already valid, there's no need to attempt to activate the machine again.

Regardless — you can list the license's current machines if you need to find a particular machine. Or yes, you can  also lookup a machine by its fingerprint when authenticated as the license:

```
GET /v1/accounts/<account>/machines/<fingerprint>
```

And whether or not a fingerprint can be shared across licenses is determined by the policy's [fingerprint uniqueness strategy](https://keygen.sh/docs/api/policies/#policies-object-attrs-fingerprintUniquenessStrategy).

Lastly, the `meta.id` is a reference ID. It is the request log's ID, so you/we can lookup the request that caused the error.

**Reply by nschloe** [reply]
> If it's already valid, there's no need to attempt to activate the machine again.

I see! This leads back to a question I had earlier this week. Some licenses _do_ require activation, others _don't_. I don't know this information before asking the API. Does the flow need to look like the following?

```mermaid
flowchart TD
    B{Valid \nw/o fingerprint?}
    B -->|success| C{Validation\nw/ fingerprint?}
    C -->|success| D[Valid!]
    C -->|fail| E{Activate machine?}
    E -->|success| F[Valid!]
    E -->|cannot activate| G[Valid!]
    E -->|other failures| H[Invalid!]
    B -->|fail| Z[Invalid!]
```

It's also possible that I'll have to rethink the parallel _with_/_without activation_ license handling.

**Reply by ezekg** [reply]
Licenses that *do* require a fingerprint will fail validation with a `FINGERPRINT_SCOPE_REQUIRED` code when attempting to validate without a fingerprint (assuming their policy requires a fingerprint scope during validation). If a license is valid without a fingerprint, no further validation nor activation is needed.

Are you writing an generalized SDK? What's the use case here?

**Reply by nschloe** [reply]
Aha, so I'd need to be more like this:

```mermaid
flowchart TD
    B{Valid \nw/o fingerprint?}
    B -->|success| C[Valid!]
    B -->|FINGERPRINT_SCOPE_REQUIRED|B2{Valid\nw/ fingerprint?}
    B -->|other fails| B3[Invalid!]
    B2 -->|success| D[Valid!]
    B2 -->|fail| E{Activate\nmachine}
    E -->|success| F[Valid!]
    E -->|fail| G[Invalid!]
```

> Licenses that do require a fingerprint

Can I distinguish those from the web UI?

> Are you writing an generalized SDK?

I'm using a license checker for multiple packages and clients, in that sense it's general, yes.

**Reply by ezekg** [reply]
That's looks correct now. But rather than having your SDK perform 2 validations, why not have a config on the SDK to enable/disable activation? That'd save on the superfluous validation requests here, and overall it may make your SDK's integration easier to reason about. Regardless — yes, the above should work.

From the UI, you'd need to look at the license's policy (either the policy object, or add something to the policy's name).

**Reply by nschloe** [reply]
> why not have a config on the SDK to enable/disable activation? 

That would be lovely. For this work, I'd have to handle the information "with/without activation" along with the key itself. If it's stored in a file, I could just store both of these things; if it's given in an env variable, I could perhaps `MY_KEY="w/<the actual key>"` and `MY_KEY="wo/<the actual key"` or something like that. ... I'm not too happy with the latter.

> From the UI, you'd need to look at the license's policy

`Scope requirements`, ah, I hadn't seen this!

**Reply by nschloe** [reply]
@ezekg Zeke, how do I configure a policy to reject fingerprinted validation requests with `FINGERPRINT_SCOPE_MISMATCH`? I just added a new policy, everything default ("Require fingerprint" off), and I get `NO_MACHINE`.

**Reply by nschloe** [reply]
This is weird. When fingerprint-validating a non-fingerprinted license, I get `NO_MACHINE` (just like I would for a fingerprinted license), and when I try to activative it, I get `LICENSE_NOT_ALLOWED: License key authentication is not allowed by policy`. What am I doing wrong?

**Reply by nschloe** [reply]
Here's a little survey of possible outcomes:

- License with no required fingerprinting:
  - validate with fingerprint:
    - err, NO_MACHINE
  - validate without fingerprint:
    - VALID
  - activate:
    - ~~err, LICENSE_NOT_ALLOWED
      (License key authentication is not allowed by policy)~~
      _Edit:_ Success!

- License which DOES require fingerprinting:
  - validate with fingerprint:
    - VALID
    - err, NO_MACHINE
  - validate without fingerprint:
    - err, FINGERPRINT_SCOPE_REQUIRED
  - activate:
    - success
    - err, MACHINE_LIMIT_EXCEEDED
    - err, FINGERPRINT_TAKEN

At validation with fingerprint, I wish there was a way to tell apart those licenses in which the machine is missing (NO_MACHINE) and those where in principle no machine can be added. The latter case is only revealed on activation now. I don't know now how to produce `FINGERPRINT_SCOPE_MISMATCH`.

**Reply by ezekg** [reply]
@nschloe there is no such thing as a non-fingerprinted license. Any license can have activations, even if the policy does not necessarily require it. If you include a fingerprint scope, the validation request will assert that scope — no matter what. There is no way to reject scopes from the policy. It's very simple: if you don't want a validation to use a scope, don't send one. Your application or SDK will need to know when and when not to send a fingerprint.

If the license has no machines, a `NO_MACHINE(S)` failure code will be returned (plurality depending on if the policy is floating). If the machine does have machines, but none of the fingerprints match, a `FINGERPRINT_SCOPE_MISMATCH` failure code is returned. Ultimately, if the policy is [strict](https://keygen.sh/docs/api/policies/#policies-object-attrs-strict), then an activation will *always* be required.

More information on the various codes are available [here](https://keygen.sh/docs/api/licenses/#licenses-licenses-actions-validate-key-response-codes-information).

Lastly — the `LICENSE_NOT_ALLOWED` error code indicates that the policy does not allow [license key authentication](https://keygen.sh/docs/api/authentication/#license-authentication). You need to set your policy's [authentication strategy](https://keygen.sh/docs/api/policies/#policies-object-attrs-authenticationStrategy) to `LICENSE` or `MIXED`.

**Reply by nschloe** [reply]
Thanks for the reply! I'll dive back into it tomorrow.

**Reply by nschloe** [reply]
Okay, I think I got it now:

- You can do license activation, validation with fingerprints etc no matter if the license has _Require fingerprint scope_ or not. 
- The demo account failed my fingerprinted validation request (https://github.com/keygen-sh/community/discussions/63) not because it in principle doesn't support it. It failed because I provided a fingerprint, and that fingerprint isn't in the list of activated machines for that license (`FINGERPRINT_SCOPE_MISMATCH`). If the demo license had had no activated licenses at all, the error code would've been `NO_MACHINE`/`NO_MACHINES`.

I could in principle share code paths between licenses that _do_ require fingerprinting and those that don't. Just always activate and validate with fingerprint.

The cleaner approach would certainly be to store the info about fingerprint requirement somewhere with the key, e.g., with a prefix.

**Reply by ezekg** [reply]
That is correct. 👍 

You could also use `metadata` on the license to determine whether or not to activate, since that information would be available after validation and before activation, via the validation response body.
